contracts/Relay.sol

```diff
@@ -0,0 +1,284 @@
+        bytes32 benefit
+    ) public whenNotPaused {
+        // verify hash, signatures (The number of signers must be greater than 2/3 of the total)
+        require(
+            _checkSignature(hash, message, signatures),
+            "Relay: Bad relayer signature"
+        );
+
+        // decode message, check nonce and relayer
+        Input.Data memory data = Input.from(message);
+        (bytes memory prefix, uint32 nonce, address[] memory authorities) = Scale.decodeAuthorities(
+            data
+        );
+
+        require(_checkNetworkPrefix(prefix), "Relay: Bad network prefix");
+
+        // update nonce,relayer
+        _setRelayer(nonce, authorities, benefit);
+    }
+
+    function appendRoot(
+        bytes32 hash,
+        bytes memory message,
+        bytes[] memory signatures
+    ) public whenNotPaused {
+        // verify hash, signatures
+        require(
+            _checkSignature(hash, message, signatures),
+            "Relay: Bad relayer signature"
+        );
+
+        // decode message, check nonce and relayer
+        Input.Data memory data = Input.from(message);
+        (bytes memory prefix, uint32 index, bytes32 root) = Scale.decodeMMRRoot(data);
+
+        require(_checkNetworkPrefix(prefix), "Relay: Bad network prefix");
+
+        // append index, root
+        _appendRoot(index, root);
+    }
+
+    function resetRoot(uint32 index, bytes32 root) public onlyOwner {
+        _setRoot(index, root);
+        emit ResetRootEvent(_msgSender(), root, index);
+    }
+
+    function verifyAndDecodeReceipt(
+        bytes32 root,
+        uint32 MMRIndex,
+        uint32 blockNumber,
+        bytes memory blockHeader,
+        bytes32[] memory peaks,
+        bytes32[] memory siblings,
+        bytes memory proofstr,
+        bytes memory key
+    ) public view whenNotPaused returns (bytes memory){
+        // verify block proof
+        require(
+            verifyBlockProof(root, MMRIndex, blockNumber, blockHeader, peaks, siblings),
+            "Relay: Block header proof varification failed"
+        );
+
+        // get state root
+        bytes32 stateRoot = Scale.decodeStateRootFromBlockHeader(blockHeader);
+
+        return getLockTokenReceipt(stateRoot, proofstr, key);
+    }
+
+    function verifyBlockProof(
+        bytes32 root,
+        uint32 MMRIndex,
+        uint32 blockNumber,
+        bytes memory blockHeader,
+        bytes32[] memory peaks,
+        bytes32[] memory siblings
+    ) public view whenNotPaused returns (bool) {
+        require(
+            _getMMRRoot(MMRIndex) != bytes32(0),
+            "Relay: Not registered under this index"
+        );
+        require(
+            _getMMRRoot(MMRIndex) == root,
+            "Relay: Root is different from the root pool"
+        );
+
+        return MMR.inclusionProof(root, MMRIndex + 1, blockNumber + 1, blockHeader, peaks, siblings);
+    }
+
+    function getLockTokenReceipt(bytes32 root, bytes memory proofstr, bytes memory key)
```

Comment:
Didn't verify events?

Reply:
If verify the event proof, the user must provide the original events, and check the events obtained from the decoding proof, so that more data (original events) can be submitted.
The current method is to search for events in the proof. If the result is empty, it means that the verification failed.